File: src/core/heartbeat_client.py
import asyncio
import logging

import httpx

from config.logging_config import setup_logging
from contracts.backend import Backend
from core.metrics_manager import MetricsManager

setup_logging()
logger = logging.getLogger(__name__)


class HeartbeatClient:
    """
    Client for sending periodic heartbeat signals to the proxy to register backend health and metrics.
    """

    def __init__(
        self,
        backend: Backend,
        proxy_url,
        heartbeat_interval,
        metrics_manager: MetricsManager,
    ):
        """
        Initialize the HeartbeatClient.

        Args:
            backend (Backend): The backend instance to report.
            proxy_url (str): The proxy URL to send heartbeats to.
            heartbeat_interval (int): Interval in seconds between heartbeats.
            metrics_manager (MetricsManager): Metrics manager for backend stats.
        """
        self.backend = backend
        self.proxy_url = proxy_url
        self.heartbeat_interval = heartbeat_interval
        self.metrics_manager = metrics_manager
        self._task = None
        self._running = False
        logger.info(f"HeartbeatClient initialized for backend {self.backend.url}")

    async def start(self):
        """
        Start the heartbeat loop as an asynchronous task.
        """
        self._running = True
        self._task = asyncio.create_task(self._heartbeat_loop())
        logger.info("Heartbeat loop started.")

    async def stop(self):
        """
        Stop the heartbeat loop and cancel the running task.
        """
        self._running = False
        if self._task:
            self._task.cancel()
        logger.info("Heartbeat loop stopped.")

    async def _heartbeat_loop(self):
        """
        Internal loop that sends heartbeat data to the proxy at regular intervals.
        """
        async with httpx.AsyncClient() as client:
            while self._running:
                try:
                    self.backend.rif_avg_latency = (
                        self.metrics_manager.get_rif_avg_latency()
                    )
                    self.backend.in_flight_requests = (
                        self.metrics_manager.get_in_flight()
                    )
                    self.backend.overall_avg_latency = (
                        self.metrics_manager.get_overall_avg_latency()
                    )
                    resp = await client.post(
                        f"{self.proxy_url}/register", json=self.backend.model_dump()
                    )
                    if resp.status_code == 200:
                        logger.info(
                            f"[Heartbeat] Registered with proxy at {self.proxy_url} as {self.backend.url} with {self.backend.model_dump()}"
                        )
                    else:
                        logger.warning(
                            f"[Heartbeat] Failed to register with proxy: {resp.text}"
                        )
                except Exception as e:
                    logger.error(f"[Heartbeat] Error registering with proxy: {e}")
                await asyncio.sleep(self.heartbeat_interval)

File: go/core/heartbeat_client.go
package core

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"openprequal/contracts"
	"time"
)

type HeartbeatClient struct {
	backend  *contracts.Backend
	proxyURL string
	interval time.Duration
	metrics  *MetricsManager
	cancel   context.CancelFunc
	client   *http.Client
}

func NewHeartbeatClient(b *contracts.Backend, proxyURL string, intervalSec int, metrics *MetricsManager) *HeartbeatClient {
	return &HeartbeatClient{backend: b, proxyURL: proxyURL, interval: time.Duration(intervalSec) * time.Second, metrics: metrics, client: &http.Client{Timeout: 5 * time.Second}}
}

// helper to post backend info to proxy register/unregister endpoints
func (h *HeartbeatClient) postToProxy(path string, payload interface{}) error {
	if h.proxyURL == "" {
		return nil
	}
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	resp, err := h.client.Post(h.proxyURL+path, "application/json", bytes.NewReader(data))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	return nil
}

func (h *HeartbeatClient) Start() {
	ctx, cancel := context.WithCancel(context.Background())
	h.cancel = cancel

	// initial register: retry a few times while proxy may be starting
	if h.backend != nil {
		maxRetries := 5
		for i := 0; i < maxRetries; i++ {
			if err := h.postToProxy("/register", h.backend); err == nil {
				break
			}
			// exponential-ish backoff
			time.Sleep(200 * time.Millisecond * time.Duration(i+1))
		}
	}

	go func() {
		ticker := time.NewTicker(h.interval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				// update backend info from metrics
				if h.backend != nil {
					h.backend.AvgLatency = h.metrics.GetAvgLatency()
					h.backend.InFlightRequests = float64(h.metrics.GetInFlight())
					// post updated info to proxy
					h.postToProxy("/register", h.backend)
				}
			case <-ctx.Done():
				return
			}
		}
	}()
}

func (h *HeartbeatClient) Stop() {
	if h.cancel != nil {
		h.cancel()
	}
	// best-effort unregister
	if h.backend != nil {
		// copy minimal struct value for unregister
		var b contracts.Backend = *h.backend
		h.postToProxy("/unregister", b)
	}
}
